cpp/json_schema_converter.cc

```diff
@@ -2790,18 +2887,268 @@ Result<JSONSchemaConverter::ObjectSpec, SchemaError> JSONSchemaConverter::ParseO
+      std::string json_pointer_ebnf = RegexToEBNF(json_pointer_regex_pattern, false);
+      string_spec.pattern = json_pointer_ebnf;
+      return ResultOk(string_spec);
+    }
+    if (format == "relative-json-pointer") {
+      // refer to draft-handrews-relative-json-pointer-01, section 3
+      std::string relative_json_pointer_regex_pattern =
+          "^(0|[1-9][0-9]*)(#|(/([\\x00-\\x2E]|[\\x30-\\x7D]|[\\x7F-\\U0010FFFF]|~[01])*)*)$";
+      std::string relative_json_pointer_ebnf =
+          RegexToEBNF(relative_json_pointer_regex_pattern, false);
+      string_spec.pattern = relative_json_pointer_ebnf;
+      return ResultOk(string_spec);
+    }
+  }
+  if (schema.count("pattern")) {
+    StringSpec string_spec;
+    if (json_format == JSONFormat::kJSON) {
+      string_spec.wrapper.first = "\\\"";
+      string_spec.wrapper.second = "\\\"";
+    }
+    if (schema.count("minLength") || schema.count("maxLength") || schema.count("format")) {
+      XGRAMMAR_LOG(WARNING) << "Specifying pattern and minLength/maxLength/format is not "
+                            << "supported yet, ignoring minLength/maxLength/format";
+    }
+    std::string regex_pattern = schema.at("pattern").get<std::string>();
+    std::string converted_regex = RegexToEBNF(regex_pattern, false);
+    string_spec.pattern = converted_regex;
+    return ResultOk(string_spec);
+  }
+  if (schema.count("minLength") || schema.count("maxLength")) {
+    StringSpec string_spec;
+    if (json_format == JSONFormat::kJSON) {
+      string_spec.wrapper.first = "\\\"";
+      string_spec.wrapper.second = "\\\"";
+    }
+    string_spec.min_length = schema.count("minLength") ? schema.at("minLength").get<int64_t>() : 0;
+    string_spec.max_length = schema.count("maxLength") ? schema.at("maxLength").get<int64_t>() : -1;
+    XGRAMMAR_CHECK(string_spec.max_length == -1 || string_spec.min_length <= string_spec.max_length)
+        << "In string schema, minLength " << string_spec.min_length << " is greater than "
+        << "maxLength " << string_spec.max_length;
+    switch (json_format) {
+      case JSONFormat::kJSON: {
+        string_spec.pattern = "[^\"\\\\\\r\\n]";
+        break;
+      }
+      case JSONFormat::kXML: {
+        string_spec.pattern = "[^<>&\\r\\n]";
+        break;
+      }
+    }
+    return ResultOk(string_spec);
+  }
+  StringSpec string_spec;
+  switch (json_format) {
+    case JSONFormat::kJSON: {
+      string_spec.pattern = "[\"] " + kBasicStringSub;
+      return ResultOk(string_spec);
+    }
+    case JSONFormat::kXML: {
+      string_spec.pattern = kXMLString;
+      return ResultOk(string_spec);
+    }
+    default: {
+      XGRAMMAR_LOG(FATAL) << "Unsupported string escape type: " << static_cast<int>(json_format);
+    }
+  }
+}
+
 std::string JSONSchemaConverter::VisitObject(
     const picojson::object& schema, const std::string& rule_name
 ) {
   // Parse the object schema
+  NestedLayerHelper helper;
```

Comment:
ObjectSpec{JSONFormat format;};

// In converter.Convert
if (kXML) {
    // Do checking: schema must be object
         return SchemaError;
    return VisitXMLObject;
}
else {
  // Do normal conversion process
}

// In VisitXMLObject
ObjectSpec ... = parse();
spec.format = kXML;
Do conversion to bnf logic...